ssh: fix relogin with jumphosts

awly opened the pull request:
Several fixes to make `tsh ssh -J leaf.proxy.com` work if the root cert
is missing/expired.

Commit:
ssh: fix relogin with jumphosts

Several fixes to make `tsh ssh -J leaf.proxy.com` work if the root cert
is missing/expired.

awly commented:
@andrejtokarcik after a bit of tweaking, this change makes the following work:
```
# simulate expired session
$ tsh login --proxy=root.example.com --insecure
...
$ rm ~/.tsh/keys/root.example.com/awly-ssh/*

$ tsh ssh --insecure -J leaf.example.com:3023 leaf-node
```

I'm not confident that all of the changes are safe though, WDYT?

andrejtokarcik commented on lib/client/keystore.go:
```diff
@@ -305,6 +305,7 @@ func (fs *FSLocalKeyStore) GetKey(idx KeyIndex, opts ...CertOption) (*Key, error
 		Priv:     priv,
 		TLSCert:  tlsCert,
 		TrustedCA: []auth.TrustedCerts{{
+			ClusterName:     idx.ClusterName,
```
I don't think this is entirely correct. `TrustedCerts.ClusterName` should correspond to the cluster from which the host CA certs were obtained (which should be the root cluster) but `idx.ClusterName` is often set to a leaf cluster's name.

awly replied:
good catch!
added a loop to properly split `tlsCA` into individual `auth.TrusterCert` based on the cluster name from the CN.

awly replied:
Removed this change for now, turns out it's not needed.

andrejtokarcik commented on lib/client/api.go:
```diff
@@ -1060,13 +1072,24 @@ func (tc *TeleportClient) LoadKeyForCluster(clusterName string) error {
 func (tc *TeleportClient) LoadKeyForClusterWithReissue(ctx context.Context, clusterName string) error {
 	err := tc.LoadKeyForCluster(clusterName)
 	if err == nil {
+		// TODO(awly): is this always correct?
+		//
+		// Reset client AuthMethods, in case they were populated via relogin
+		// and would force us to use the root cert.
+		if len(tc.AuthMethods) > 0 {
+			tc.AuthMethods = nil
+		}
```
I don't think this is needed at all. `tc.AuthMethods` (= `tc.Config.AuthMethods`) is never populated via relogin. It's always set explicitly alongside with `SkipLocalAuth = true`, i.e. `tc.AuthMethods` is never really involved in typical login scenarios. It's used either from `tsh` (when using an identity file) or the web terminal client.

awly replied:
It's populated in https://github.com/gravitational/teleport/blob/fbae7ad508d0f30aacbfb2bf5a3878e1c448439b/lib/client/api.go#L472-L481

Here's what happens when relogin triggers in `RetryWithRelogin`:
- user logs in
- root cluster cert is issued
- root cluster cert is set in `tc.Config.AuthMethods`
- on next attempt, `ConnectToProxy` uses `tc.Config.AuthMethods[0]` first
- this cert is accepted by the leaf proxy, because CA validation succeeds
- the leaf proxy ends up routing the connection to `node@root` instead of `node@leaf` because of the cert cluster name extension
- the node connection fails because `leaf` can't forward requests to `root`

Because node connection happens after a successful `ConnectToProxy`, the client doesn't try any other `tc.authMethod`s.

andrejtokarcik replied:
Right I didn't notice this occurrence.

This assignment to `tc.Config.AuthMethods` in `RetryWithRelogin` is just an optimization? Wouldn't it be easier to drop this assignment, and rely on the agent auth methods only...?

awly replied:
You're right.
I was afraid to remove it, assuming something relies on the AuthMethods update.
But removing it doesn't seem to break anything on relogin (based on manual testing), so let's do that.

Worst case, we should catch any issues during the testplan.

andrejtokarcik commented on lib/client/api.go:
```diff
@@ -462,7 +462,19 @@ func RetryWithRelogin(ctx context.Context, tc *TeleportClient, fn func() error)
 	}
 	if err := tc.ActivateKey(ctx, key); err != nil {
-		return trace.Wrap(err)
+		if len(tc.JumpHosts) == 0 {
+			return trace.Wrap(err)
+		}
+		errViaJumphost := err
+		// ActivateKey re-fetches the list of CAs from auth server. If
+		// JumpHosts was pointing at the leaf cluster, this could've caused the
+		// above error. Try to ActivateKey without JumpHosts to force it to use
+		// the root cluster.
+		if err := tc.WithoutJumpHosts(func(tc *TeleportClient) error {
+			return tc.ActivateKey(ctx, key)
+		}); err != nil {
+			return trace.NewAggregate(errViaJumphost, err)
+		}
```
Even though this should work, I don't like this general approach. `WithoutJumpHosts` is wonky, we should focus on when jumphosts *are* supposed to be used, i.e. take a `WithJumpHosts` approach.

Also if we're going down this way, this should probably be put directly into `TeleportClient.UpdateTrustedCA`, not ad hoc in `RetryWithRelogin`.

awly replied:
yeah, it's pretty ugly.
the whole client library is a spaghetti and I don't know where jumphosts really matter and where not.
moved the retry into `ActivateKey` for now.

Commit:
Address review feedback

Correctly parse trusted CAs on GetKey.
Move retry without jumphosts from relogin to UpdateClusterCAs.

Commit:
Remove TelpoertClient.AuthMethods override on relogin

It doesn't seem to break anything.

r0mant commented on lib/client/api.go:
```diff
@@ -469,15 +469,6 @@ func RetryWithRelogin(ctx context.Context, tc *TeleportClient, fn func() error)
 		log.Warningf("Failed to save profile: %v", err)
 		return trace.Wrap(err)
 	}
-	// Override client's auth methods, current cluster and user name
-	authMethod, err := key.AsAuthMethod()
-	if err != nil {
-		return trace.Wrap(err)
-	}
-	// After successful login we have local agent updated with latest
-	// and greatest auth information, setup client to try only this new
-	// method fetched from key, to isolate the retry
-	tc.Config.AuthMethods = []ssh.AuthMethod{authMethod}
```
Why did you remove this?

awly replied:
This messes up the relogin flow.

`tc.Config.AuthMethods` is used as the first choice when connecting to a proxy over SSH.
If you were originally connecting to a leaf cluster (as in `tsh ssh -J leaf.proxy.com somenode`) and relogin kicked in, you're be logged in via the root cluster (which is correct).
But after you've re-logged through the root cluster `tc.Config.AuthMethods` would have an SSH cert for the root cluster, not for the leaf.
The cert caching and loading logic that @andrejtokarcik added correctly adds a leaf cert into the `localAgent`, but it gets overriden by `tc.Config.AuthMethods`.

r0mant reviewed (approved):
Code lgtm but a question about the removed block.

russjones reviewed (approved):
Bot.

Commit:
Merge branch 'master' into andrew/jumphost-relogin-fix
